pilot/pkg/serviceregistry/federation/controller.go

```diff
@@ -1020,27 +1033,14 @@ func (c *Controller) watch(eventCh chan *federationmodel.WatchEvent, stopCh <-ch
 	dec := json.NewDecoder(resp.Body)
 	defer resp.Body.Close()
 	for {
-		reader := dec.Buffered().(*bytes.Reader)
```

Comment:
whoops. I added this part in my last PR to make sure not to leave any goroutines around when stopping a watch. I have tested it extensively, though and I'm surprised you say it will never work

Reply:
maybe I was testing an old image?

Reply:
Hm, yeah, using an old image sounds like the most likely explanation. 

If you check the decoder code in golang/src/encoding/json/stream.go, you'll see that the buffer is only written to in the `Decode()` function.